linera-chain/src/certificate/hashed.rs

```diff
@@ -102,6 +108,60 @@ impl<T> PartialEq for Hashed<T> {
 impl<T> Eq for Hashed<T> {}
 
 /// A constraint summing a value of type `T`.
-pub trait Has<T> {
-    fn get(&self) -> &T;
+pub trait Has<T, R = T> {
+    fn get(&self) -> R;
+}
+
+pub enum IsValidated {}
+
+impl Has<IsValidated, bool> for ValidatedBlock {
+    fn get(&self) -> bool {
+        true
+    }
+}
```

Comment:
This `Has` trait is starting to look very suspicious.

Reply:
Why? I think it's great (although I'm biased of course) way to ensure that a type `T` has some specific getter.

Reply:
I don't understand what difference does it make 🤔 it many places the values we get out of the getters are used only before they're signed. And once they're part of the certificate, the getters are not used anymore.

Reply:
And actually I think the previous way it was done was incorrect - the `CertificateValue` was a "catch-it-all" struct that included chainid, round, epoch but also required blob ids, whether value was of a certain type (`is_validated`). It wasn't "one thing", it was many things.

Reply:
I'd say it makes at least a difference in terms of how understandable the code is: "This type can end up as the value of a certificate" tells me more than "this type has a chain ID".

Reply:
I'd say that the point of working with generic functions is to require as little as possible from the caller. If I were to replace this approach with something like a `trait CertificateValueT` (working name) and require that instead then some places will require unnecessary constraints from the arguments and the implementations for these types will have dummy values that don't make sense (just like it was before).

I'll do the change and we'll judge again.

Reply:
If we end up in that situation it just means we applied the `CertificateValueT` trait to something that actually isn't a certificate value, just to get a chain ID getter.

Are there really any examples in the code where we handle "anything with a chain ID" but it's not necessarily a certificate value?

If we do, I'd find that suspicious, but if we really need it, we could always `impl<T> Has<ChainId> for T where T: CertificateValueT`.

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2921/commits/db80ea5d48dfef41494414d91dcff47e1b2cd1ab . I think you might have been right and it does _read_ better.

Reply:
Just FYI @ma2bd  - I'm resolving this comment as `Has` trait has been removed entirely from the codebase.